packages/origin-247-certificate/src/evented-certificate/certificate.facade.ts

```diff
@@ -0,0 +1,167 @@
+import { Injectable, Inject } from '@nestjs/common';
+import {
+    CERTIFICATE_COMMAND_REPOSITORY,
+    CertificateCommandRepository
+} from './repositories/CertificateCommand/CertificateCommand.repository';
+import {
+    CERTIFICATE_EVENT_REPOSITORY,
+    CertificateEventRepository
+} from './repositories/CertificateEvent/CertificateEvent.repository';
+import { CertificateService } from '../certificate.service';
+import { QueryBus, EventBus } from '@nestjs/cqrs';
+import { GetAllCertificatesQuery, GetCertificateQuery } from '@energyweb/issuer-api';
+import { CertificateEventType } from './events/Certificate.events';
+
+import {
+    ICertificate,
+    IClaimCommand,
+    IIssueCommand,
+    ITransferCommand,
+    ISuccessResponse,
+    BlockchainActionType,
+    BlockchainAction,
+    IIssuedCertificate,
+    IIssueCommandParams
+} from '../types';
+import {
+    CertificateIssuedEvent,
+    CertificateTransferredEvent,
+    CertificateClaimedEvent
+} from './events/Certificate.events';
+
+const CertificateEventVersion = 1;
+
+type CertificateCommand = IIssueCommand<unknown> | ITransferCommand | IClaimCommand;
+@Injectable()
+export class CertificateFacade<T = null> {
+    constructor(
+        @Inject(CERTIFICATE_COMMAND_REPOSITORY)
+        private readonly certCommandRepo: CertificateCommandRepository,
+        @Inject(CERTIFICATE_EVENT_REPOSITORY)
+        private readonly certEventRepo: CertificateEventRepository,
+        private readonly certService: CertificateService,
+        private readonly queryBus: QueryBus,
+        private readonly eventBus: EventBus
+    ) {}
+
+    public async issue(params: IIssueCommandParams<T>): Promise<IIssuedCertificate<T>> {
+        const command = {
+            ...params,
+            fromTime: Math.round(params.fromTime.getTime() / 1000),
+            toTime: Math.round(params.toTime.getTime() / 1000)
+        } as IIssueCommand<T>;
+
+        const savedCommand = await this.certCommandRepo.create({ payload: command });
+        this.validateCommand(command); // this will probably throw
+
+        const internalCertId = this.generateInternalCertificateId();
+        const event = new CertificateIssuedEvent(internalCertId, CertificateEventVersion, command);
+
+        this.eventBus.publish(event);
+        this.certEventRepo.create({
```

Comment:
To be honest, I would to it like that: `this.certEventRepo.save(event)` and let repository handle event mapping

Reply:
It cannot just take event, since we decided to have the command that caused the event.

Reply:
I don't understand. Just send event to the repository, and this will do the mapping to database representation. You can pass commandId as additional parameter.

Reply:
Added save method that takes event and commandId.